### PR TITLE
Batch analysis from mongodb operator

### PR DESCRIPTION
**Pull request by laphets**
## Result

Analyze `disable-enable-shard` workload ([ref](https://github.com/sieve-project/sieve-issue-only/issues/57)) from mongodb operator, and got the following result initially,

- 127 configs generated from the learning mode in total
- 2 statefulset/default/mongodb-cluster-cfg DELETE inconsistency
- 4 DELETE inconsistency
  - `statefulset/default/mongodb-cluster-cfg`
  - `service/default/mongodb-cluster-cfg`
  - `deployment/default/mongodb-cluster-mongos`
- 62 CREATE inconsistency
  - `service/default/mongodb-cluster-mongos`
  - `deployment/default/mongodb-cluster-mongos`
  - `statefulset/default/mongodb-cluster-cfg`

And only 2 / 127 have successfully matched for the crucial event and the reverse side effect event (analyzed from the server log).

## Features
- Add `sleep over` checking before bug report generating, do not generate the bug report if sieve server didn't successfully match for the crucial event and reverse side effect event.
- Reducing config by check whether a reverse side effect event appear after the side effect happening under the learning mode (127 -> 65)

## Bug Fix
- Random node selection for resource (by scheduler) -> make the `nodeName` field as placeholder in crucial event
- Random IP assign for pod -> make `ip*` field as placeholder in crucial event
- Fields for resource object are case sensitive, there is inconsistency between field name in crucial event and runtime event. e.g. `spec` in crucial event (get from client-go) and `Spec` in runtime event (get from API server).
- Support nil value checking inside crucial event matching func.

## TODO
- need to run the fixed version sieve on the workload for time-travel and compare the result

### CODE REVIEW

**Comment by tianyin** [comment]
This is again an awesome PR as always. Thank @laphets!

@marshtompsxd Could you take a look at the Bug Fixes?

**Comment by marshtompsxd** [comment]
Yes, I will do a careful pass. We have discussed the results and bug fixing in this PR.

File: analyze_event.py [hunk 1]
```diff
@@ -102,25 +102,30 @@ def diff_events(prev_event, cur_event):
...
         elif isinstance(event[key], list):
-            canonicalize_event_for_list(event[key])
+            canonicalize_event_for_list(event[key], node_ignore)
         elif isinstance(event[key], str):
             if re.match(common.TIME_REG, str(event[key])):
                 event[key] = common.SONAR_CANONICALIZATION_MARKER
+            if node_ignore:
+                if 'ip' in key.lower() and re.match(common.IP_REG, str(event[key])):
+                    event[key] = common.SONAR_CANONICALIZATION_MARKER
+                if key in ["nodeName", "containerID"]:
```
Comment (marshtompsxd) [inline analyze_event.py:129]: We can def this list in common.py (and maybe later make it configurable)

File: analyze.py [hunk 2]
```diff
@@ -388,8 +388,24 @@ def dump_json_file(dir, data, json_file_name):
...
+    filtered_causality_pairs = []
+    for pair in causality_pairs:
+        start_event = pair[0]
+        side_effect = pair[1] # e.g. delete sth
+        flag = False
+        for event in event_key_map[side_effect.key]:
+            # We will find add sth
+            if event.start_timestamp <= side_effect.end_timestamp:
+                continue
+            if event.etype == "Added":
```
Comment (marshtompsxd) [inline analyze.py:401]: You might also want to check whether the side effect is a delete or add (tho currently we are only considering delete)

File: controllers.py [hunk 3]
```diff
@@ -86,7 +88,7 @@ def __init__(self, workload, config, mode, two_sided=False, num_workers=2):
         "recreate": Suite(
             workloads.workloads["mongodb-operator"]["recreate"], "test-mongodb-operator/test/time-travel-1.yaml", "time-travel", num_workers=3),
         "disable-enable-shard": Suite(
-            workloads.workloads["mongodb-operator"]["disable-enable-shard"], "test-mongodb-operator/test/time-travel-2.yaml", "time-travel", num_workers=3),
+            workloads.workloads["mongodb-operator"]["disable-enable-shard"], "test-mongodb-operator/test/time-travel-2.yaml", "time-travel", num_workers=3, se_filter=True),
```
Comment (marshtompsxd) [inline controllers.py:91]: actually, I think the se filter should be turned on by default

File: sonar-server/time-travel-server.go [hunk 4]
```diff
@@ -82,12 +84,56 @@ func (s *timeTravelServer) Start() {
...
 }
 
+func toLowerMap(m map[string]interface{}) {
+	for key, val := range m {
+		switch v := val.(type) {
+		case map[string]interface{}:
+			if e, ok := m[key].(map[string]interface{}); ok {
+				toLowerMap(e)
+				m[strings.ToLower(key)] = e
+				if strings.ToLower(key) != key {
```
Comment (marshtompsxd) [inline sonar-server/time-travel-server.go:94]: this is minor, but you might want to put `m[strings.ToLower(key)] = e` inside the if scope. Similar for the `m[strings.ToLower(key)] = e` below

File: sonar-server/time-travel-server.go [hunk 5]
```diff
@@ -82,12 +84,56 @@ func (s *timeTravelServer) Start() {
...
+			} else {
+				log.Println("bad")
+			}
+
+		case []interface{}:
+			if e, ok := m[key].([]interface{}); ok {
+				for idx := range e {
+					switch e[idx].(type) {
+					case map[string]interface{}:
+						if ee, ok := e[idx].(map[string]interface{}); ok {
```
Comment (marshtompsxd) [inline sonar-server/time-travel-server.go:106]: try to find a better name than ee :)

File: sonar-server/time-travel-server.go [hunk 6]
```diff
@@ -82,12 +84,56 @@ func (s *timeTravelServer) Start() {
...
+							e[idx] = ee
+						}
+					}
+				}
+				m[strings.ToLower(key)] = e
+				if strings.ToLower(key) != key {
+					delete(m, key)
+				}
+			} else {
+				log.Println("bad")
```
Comment (marshtompsxd) [inline sonar-server/time-travel-server.go:117]: You can make the log message clearer (which also makes your life easier)

File: sonar-server/time-travel-server.go [hunk 7]
```diff
@@ -82,12 +84,56 @@ func (s *timeTravelServer) Start() {
...
+			}
+
+		case []interface{}:
+			if e, ok := m[key].([]interface{}); ok {
+				for idx := range e {
+					switch e[idx].(type) {
+					case map[string]interface{}:
```
Comment (marshtompsxd) [inline sonar-server/time-travel-server.go:105]: You might also need to consider the case where you have list in list `case []interface{}:`.
If you do not want to handle list in list now, just leave an error message and exit in that case

File: sonar-server/time-travel-server.go [hunk 8]
```diff
@@ -82,12 +84,56 @@ func (s *timeTravelServer) Start() {
...
+		switch v := val.(type) {
+		case map[string]interface{}:
+			if e, ok := m[key].(map[string]interface{}); ok {
+				toLowerMap(e)
+				m[strings.ToLower(key)] = e
+				if strings.ToLower(key) != key {
+					delete(m, key)
+				}
+			} else {
+				log.Println("bad")
```
Comment (marshtompsxd) [inline sonar-server/time-travel-server.go:98]: You can make the log message clearer (which also makes your life easier)

**Comment by tianyin** [comment]
> We have discussed the results and bug fixing in this PR.

Great! I was not aware of it.

**Comment by tianyin** [comment]
@laphets Feel free to merge after addressing the comments. 

Overall, LGTM and great job!

File: sonar-server/time-travel-server.go [hunk 9]
```diff
@@ -82,12 +84,56 @@ func (s *timeTravelServer) Start() {
...
+			}
+
+		case []interface{}:
+			if e, ok := m[key].([]interface{}); ok {
+				for idx := range e {
+					switch e[idx].(type) {
+					case map[string]interface{}:
+						if ee, ok := e[idx].(map[string]interface{}); ok {
+							toLowerMap(ee)
+							e[idx] = ee
```
Comment (marshtompsxd) [inline sonar-server/time-travel-server.go:108]: This part might be tricky. After `ee, ok := e[idx].(map[string]interface{})`, is ee a reference to e[idx] or a deep copy? If it is a reference then there is no need to `e[idx] = ee` I guess?
Correct me if I am wrong.

File: sonar-server/time-travel-server.go [hunk 10]
```diff
@@ -220,8 +267,16 @@ func equivalentEvent(crucialEvent, currentEvent map[string]interface{}) bool {
 			} else {
 				return false
 			}
+
 		default:
-			log.Printf("Unsupported type: %v %T\n", v, v)
+			if e, ok := currentEvent[key]; ok {
+				if val == nil && e == nil {
+					log.Printf("Both nil type: %v and %v , key: %s\n", v, e, key)
```
Comment (marshtompsxd) [inline sonar-server/time-travel-server.go:274]: So did you encounter a case where both currentEvent[key] and crucialEvent[key] are of nil type?

File: workloads.py [hunk 11]
```diff
@@ -53,7 +53,7 @@
         .cmd("kubectl apply -f test-mongodb-operator/test/cr-shard.yaml").wait_for_pod_status("mongodb-cluster-rs0-2", common.RUNNING).wait_for_pod_status("mongodb-cluster-cfg-2", common.RUNNING)
         .cmd("kubectl patch PerconaServerMongoDB mongodb-cluster --type merge -p='{\"spec\":{\"sharding\":{\"enabled\":false}}}'").wait_for_pod_status("mongodb-cluster-cfg-2", common.TERMINATED)
         .cmd("kubectl patch PerconaServerMongoDB mongodb-cluster --type merge -p='{\"spec\":{\"sharding\":{\"enabled\":true}}}'").wait_for_pod_status("mongodb-cluster-cfg-2", common.RUNNING)
-        .wait(50),
+        .wait(120),
```
Comment (marshtompsxd) [inline workloads.py:56]: I guess the 120 is for debugging,
you can change it back to 50 if the cluster can get stable after 50 seconds

**Comment by marshtompsxd** [comment]
@laphets 
Next time let's open an issue and discuss the bugs/challenges there, so we can keep the complete history (avoiding partial history!) of the important decisions/revisions we made, and others can also join the discission and better help you.

File: sonar-server/time-travel-server.go (hunk 10 above)
Comment (laphets) [reply sonar-server/time-travel-server.go:274]: Yes, for example, `"lastProbeTime": null`, a filed is null in the json

File: sonar-server/time-travel-server.go (hunk 9 above)
Comment (laphets) [reply sonar-server/time-travel-server.go:108]: I think so, since the element we assert is a map, so it's taken by reference

**Comment by tianyin** [comment]
@laphets shall we merge this?

**Comment by laphets** [comment]
> @laphets shall we merge this?

Let me finish the no `sleep over` investigation and then merge together.

**Comment by tianyin** [comment]
Awesome! @laphets !

**Comment by marshtompsxd** [comment]
@laphets Just ensure that after merging the PR the previous bugs can still be reproduced, especially the cass operator one

**Comment by tianyin** [comment]
@marshtompsxd is it ensured by passing all the test? Or it's not guaranteed?

**Comment by marshtompsxd** [comment]
The current CI will run the sieve test but I am not sure whether the CI also checks the content in the bug report.
If not, I can help implement some checkers in the CI.

**Comment by laphets** [comment]
> The current CI will run the sieve test but I am not sure whether the CI also checks the content in the bug report.
> If not, I can help implement some checkers in the CI.

Yes, I think we need some additional checking for the generated `bug-report.txt` (maybe after the sieve test).

**Comment by laphets** [comment]
> @laphets Just ensure that after merging the PR the previous bugs can still be reproduced, especially the cass operator one

Looks good in [CI](https://github.com/sieve-project/sieve/pull/5/checks?check_run_id=2845302965#step:11:142). Merged.

**Comment by tianyin** [comment]
This is a huge merge! Congratulations @laphets !!!

**Pull request merged**